alertmanager/handlers/template_handlers_test.go

```diff
@@ -25,118 +28,310 @@ var (
+			Filename: "not_a_file",
+			ExpectedError: "code=400, message=error getting file: file not_a_file does not exist",
+		},
+		{
+			Name: "template client error",
+			TmplClientExpectedReturn: []interface{}{errors.New("template error")},
+			ExpectedError: "code=500, message=error adding template: template error",
+		},
+	}
+	runAllTests(t, tests, baseTest)
+}
+
+func TestGetPutTemplateHandler(t *testing.T) {
+	baseTest := templateTestCase{
+		Name:                     "successful put",
+		Filename:                 "file1",
+		TmplClientFunc:           "EditTemplate",
+		TmplClientExpectedParams: []interface{}{mock.Anything, mock.Anything, mock.Anything},
+		TmplClientExpectedReturn: []interface{}{nil},
+		HandlerFunc:              GetPutTemplateHandler,
+	}
+	tests := []templateTestCase{
+		baseTest,
+		{
+			Name: "file doesn't exist",
+			Filename: "not_a_file",
+			ExpectedError: "code=400, message=error getting template: file not_a_file does not exist",
+		},
+		{
+			Name: "template client error",
+			TmplClientExpectedReturn: []interface{}{errors.New("template error")},
+			ExpectedError: "code=500, message=error editing template: template error",
+		},
+	}
+	runAllTests(t, tests, baseTest)
+}
+
+func TestGetDeleteTemplateHandler(t *testing.T) {
+	baseTest := templateTestCase{
+		Name:                     "successful delete",
+		Filename:                 "file1",
+		TmplClientFunc:           "DeleteTemplate",
+		TmplClientExpectedParams: []interface{}{mock.Anything, mock.Anything},
+		TmplClientExpectedReturn: []interface{}{nil},
+		HandlerFunc:              GetDeleteTemplateHandler,
+	}
+	tests := []templateTestCase{
+		baseTest,
+		{
+			Name: "file doesn't exist",
+			Filename: "not_a_file",
+			ExpectedError: "code=400, message=error getting template: file not_a_file does not exist",
+		},
+		{
+			Name: "template client error",
+			TmplClientExpectedReturn: []interface{}{errors.New("template error")},
+			ExpectedError: "code=500, message=error deleting template: template error",
+		},
+	}
+	runAllTests(t, tests, baseTest)
+}
+
+
 func getTestAMClient() *mocks.AlertmanagerClient {
 	client := mocks.AlertmanagerClient{}
 	client.On("GetTemplateFileList").Return(sampleFileList, nil)
 	return &client
 }
 
-func buildTmplContext(body interface{}, method, target, path, tmplFileName string) (echo.Context, *httptest.ResponseRecorder) {
-	bytes, _ := json.Marshal(body)
-	req := httptest.NewRequest(method, target, strings.NewReader(string(bytes)))
-	rec := httptest.NewRecorder()
-	c := echo.New().NewContext(req, rec)
-	c.SetPath(path)
-	c.Set(templateFilenameParam, tmplFileName)
-	return c, rec
+func getTestTmplClient() *mocks.TemplateClient {
+	client := mocks.TemplateClient{}
+	client.On("Root").Return(sampleRootDir)
+	return &client
+}
+
+func runAllTests(t *testing.T, tests []templateTestCase, baseTest templateTestCase) {
+	for i := range tests {
+		err := mergo.Merge(&tests[i], baseTest)
+		assert.NoError(t, err)
```

Comment:
This is going to make it more difficult to debug if a test does fail (ie, which test failed?).  Can we either not have this function, or include more context in the error?

Reply:
The way the `t.Run` function works means it'll end up looking like this: 
```
=== RUN   TestGetGetTemplateFileHandler/non-existent_file
    --- PASS: TestGetGetTemplateFileHandler/non-existent_file (0.00s)
=== RUN   TestGetGetTemplateFileHandler/template_client_error
    --- PASS: TestGetGetTemplateFileHandler/template_client_error (0.00s)
=== RUN   TestGetPostTemplateFileHandler
    --- PASS: TestGetPostTemplateFileHandler (0.00s)
```

Reply:
so we'll still know exactly which test and test case failed. Gotta love Go!